src/routers/alpha-router/functions/compute-all-routes.ts

```diff
@@ -104,9 +122,103 @@ export function computeAllRoutes<
   computeRoutes(tokenIn, tokenOut, [], poolsUsed);
 
   log.info(
-    { routes: routes.map(routeToString) },
+    {
+      routes: routes.map(routeToString),
+    },
     `Computed ${routes.length} possible routes.`
   );
 
   return routes;
 }
+
+/// separate function for computing MixedRoutes to accomodate extra logic
+export function computeMixedRoutes<
+  TPool extends Pair | Pool,
+  TRoute extends MixedRoute
+>(
+  tokenIn: Token,
+  tokenOut: Token,
+  buildRoute: (route: TPool[], tokenIn: Token, tokenOut: Token) => TRoute,
+  pools: TPool[],
+  maxHops: number
+): TRoute[] {
+  const poolsUsed = Array<boolean>(pools.length).fill(false);
+  const routes: TRoute[] = [];
+
+  const computeRoutes = (
+    tokenIn: Token,
+    tokenOut: Token,
+    currentRoute: TPool[],
+    poolsUsed: boolean[],
+    _previousTokenOut?: Token
+  ) => {
+    if (currentRoute.length > maxHops) {
+      return;
+    }
+
+    if (
+      currentRoute.length > 0 &&
+      currentRoute[currentRoute.length - 1]!.involvesToken(tokenOut)
+    ) {
+      routes.push(buildRoute([...currentRoute], tokenIn, tokenOut));
+      return;
+    }
+
+    for (let i = 0; i < pools.length; i++) {
+      if (poolsUsed[i]) {
+        continue;
+      }
+
+      const curPool = pools[i]!;
+      const previousTokenOut = _previousTokenOut ? _previousTokenOut : tokenIn;
+
+      if (!curPool.involvesToken(previousTokenOut)) {
+        continue;
+      }
+
+      const currentTokenOut = curPool.token0.equals(previousTokenOut)
+        ? curPool.token1
+        : curPool.token0;
+
+      /// prevent round tripping
+      if (
+        _.flatten(
+          currentRoute.map((pool) => [pool.token0.symbol, pool.token1.symbol])
+        ).includes(currentTokenOut.symbol)
+      ) {
+        continue;
+      }
```

Comment:
Should this be allowed for mixedRoutes? It's not reusing the same pool, but the same token between pools.

i.e.

WETH -[v3]-> DAI -[v2] -> USDC -[v3]-> WETH -[v2]-> APE